File: sdks/java/src/main/java/semacore/SemaJson.java
package semacore;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

/**
 * 事件/返回值 payload 的取值工具：JSON 级一致原则下（payload 不做类型化，字段名与 core 相同），
 * 宿主高频进行「取字段 + 判空 + 判类型」，本类把这段样板收进 SDK。
 *
 * <p>≙ Node 的解构：{@code ({ delta }) => ...} 对应 {@code SemaJson.str(data, "delta")}。
 * 所有方法对 null / 非对象 / 字段缺失 / 类型不符都返回缺省值，不抛异常。
 * 类型化 payload 由三期 manifest 代码生成提供，本类是其之前的过渡与补充。</p>
 */
public final class SemaJson {

    /** 取字符串字段；缺失/类型不符返回 null */
    public static String str(JsonElement data, String key) {
        return str(data, key, null);
    }

    public static String str(JsonElement data, String key, String fallback) {
        JsonElement v = get(data, key);
        return v != null && v.isJsonPrimitive() ? v.getAsString() : fallback;
    }

    public static boolean bool(JsonElement data, String key, boolean fallback) {
        JsonElement v = get(data, key);
        return v != null && v.isJsonPrimitive() && v.getAsJsonPrimitive().isBoolean()
                ? v.getAsBoolean() : fallback;
    }

    public static long i64(JsonElement data, String key, long fallback) {
        JsonElement v = get(data, key);
        return v != null && v.isJsonPrimitive() && v.getAsJsonPrimitive().isNumber()
                ? v.getAsLong() : fallback;
    }

    public static int i32(JsonElement data, String key, int fallback) {
        return (int) i64(data, key, fallback);
    }

    /** 取原始字段（JsonNull 视同缺失）；缺失返回 null */
    public static JsonElement get(JsonElement data, String key) {
        if (data == null || !data.isJsonObject()) return null;
        JsonObject obj = data.getAsJsonObject();
        JsonElement v = obj.get(key);
        return v == null || v.isJsonNull() ? null : v;
    }

    private SemaJson() {}
}

File: sdks/csharp/Api/SemaJson.cs
using System.Text.Json;

namespace Semacore;

/// <summary>
/// 事件 payload 的取值工具（≙ Java api/SemaJson）：事件回调 data 为 <see cref="JsonElement"/>（JSON 级一致，
/// 字段名与 core 相同），宿主高频进行「取字段 + 判空 + 判类型」，本类把这段样板收进 SDK。
///
/// <para>≙ Node 的解构：<c>({ delta }) => ...</c> 对应 <c>SemaJson.Str(data, "delta")</c>。
/// 所有方法对 null / 非对象 / 字段缺失 / 类型不符都返回缺省值，不抛异常。
/// 需要完整类型化时用 <c>Semacore.Events</c> 下对应 DTO 反序列化。</para>
/// </summary>
public static class SemaJson
{
    /// <summary>取字符串字段；缺失/类型不符返回 fallback（默认 null）。</summary>
    public static string? Str(JsonElement? data, string key, string? fallback = null)
    {
        var v = Get(data, key);
        return v switch
        {
            { ValueKind: JsonValueKind.String } e => e.GetString(),
            { ValueKind: JsonValueKind.Number or JsonValueKind.True or JsonValueKind.False } e => e.GetRawText(),
            _ => fallback,
        };
    }

    public static bool Bool(JsonElement? data, string key, bool fallback)
    {
        return Get(data, key)?.ValueKind switch
        {
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            _ => fallback,
        };
    }

    public static long I64(JsonElement? data, string key, long fallback)
    {
        if (Get(data, key) is { ValueKind: JsonValueKind.Number } e)
        {
            if (e.TryGetInt64(out var n)) return n;
            if (e.TryGetDouble(out var d)) return (long)d;
        }
        return fallback;
    }

    public static int I32(JsonElement? data, string key, int fallback)
        => (int)I64(data, key, fallback);

    /// <summary>取原始字段（JSON null 视同缺失）；缺失返回 null。</summary>
    public static JsonElement? Get(JsonElement? data, string key)
    {
        if (data is not { ValueKind: JsonValueKind.Object } obj) return null;
        if (!obj.TryGetProperty(key, out var v) || v.ValueKind == JsonValueKind.Null) return null;
        return v;
    }
}
